stan/math/prim/fun/trace.hpp

```diff
@@ -16,12 +17,12 @@ namespace math {
  * @param[in] m Specified matrix.
  * @return Trace of the matrix.
  */
-template <typename T>
-inline T trace(const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
+template <typename T, typename = require_eigen_t<T>>
+inline value_type_t<T> trace(const T& m) {
```

Comment:
Does auto deduction not work here?

Reply:
It would work. But if the type is not complex or long I prefere not to use auto.